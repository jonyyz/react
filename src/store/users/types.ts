export interface IGeoLocation {
  lat: string;
  lng: string;
}

export interface IAddress {
  street: string;
  suite?: string;
  city: string;
  zipcode: string;
  geo: IGeoLocation;
}

export interface ICompany {
  name: string;
  catchPhrase: string;
  bs: string;
}

export interface IUser {
  id: number;
  name: string;
  username: string;
  email: string;
  address: IAddress;
  phone: string;
  website: string;
  company: ICompany;
}

export interface IUserBasicInfo {
  id: number;
  name: string;
  website: string;
  city: string;
  companyName: string;
}

export interface IUsersState {
  fetching: boolean;
  fetched: boolean;
  users?: IUserBasicInfo[];
  fetchError?: string;
  actionError?: string;
}
