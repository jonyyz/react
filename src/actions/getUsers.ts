import urlJoin from "url-join";
import axios from "axios";

import store from "../store";
import {
  getUsersActionCreator,
  getUsersReceiveActionCreator,
  getUsersReceiveErrorActionCreator
} from "../store/users/actions";
import {
  IUser,
  IUserBasicInfo
} from "../store/users/types";

function getUsers() {
  return store.dispatch(
    async dispatch => {
      dispatch(getUsersActionCreator());

      try {
        const response = await axios.get(urlJoin(SERVICE_URL, "/users"));
        const users: IUserBasicInfo[] = response.data.map((user: IUser): IUserBasicInfo => {
          const {
            id,
            name,
            website,
            address: { city },
            company: {
              name: companyName
            }
          } = user;

          return {
            id,
            name,
            website,
            city,
            companyName
          };
        });

        dispatch(getUsersReceiveActionCreator(users));
      } catch (e) {
        dispatch(getUsersReceiveErrorActionCreator(e.message));
      }
    }
  );
}

export default getUsers;
