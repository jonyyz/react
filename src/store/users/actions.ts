import { IUserBasicInfo } from "./types";

export const ACTION_USERS_REQUEST          = "users:request";
export const ACTION_USERS_RECEIVE          = "users:receive";
export const ACTION_USERS_RECEIVE_ERROR    = "users:receive:error";

// Action Definitions
export interface UsersRequestAction {
  readonly type: typeof ACTION_USERS_REQUEST
}

export type UsersReceiveAction = {
  readonly type: typeof ACTION_USERS_RECEIVE,
  users: IUserBasicInfo[]
}

export type UsersReceiveErrorAction = {
  readonly type: typeof ACTION_USERS_RECEIVE_ERROR,
  message: string
}

// Union Action Types
export type UsersAction =
  UsersRequestAction |
  UsersReceiveAction |
  UsersReceiveErrorAction;

// Action Creators
export const getUsersActionCreator = (): UsersRequestAction => {
  return { type: ACTION_USERS_REQUEST };
}

export const getUsersReceiveActionCreator = (users: IUserBasicInfo[]): UsersReceiveAction => {
  return { type: ACTION_USERS_RECEIVE, users };
}

export const getUsersReceiveErrorActionCreator = (message: string): UsersReceiveErrorAction => {
  return { type: ACTION_USERS_RECEIVE_ERROR, message };
}
