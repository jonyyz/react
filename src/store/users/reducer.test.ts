import * as Actions from "./actions";
import * as Types from "./types";
import Reducer from "./reducer";

describe("Users reducer", () => {
  test("UsersReceiveAction", () => {
    const state = Reducer(null, Actions.getUsersActionCreator());
    expect(state).toMatchSnapshot();
  });

  test("UsersReceiveAction", () => {
    const users: Types.IUserBasicInfo[] = [
      {
        id: 42,
        name: "Mel Blanc",
        website: "acme.org",
        city: "Walla Walla Washington",
        companyName: "ACME"
      }
    ];
    const state = Reducer(null, Actions.getUsersReceiveActionCreator(users));
    expect(state).toMatchSnapshot();
  });

  test("UsersReceiveError", () => {
    const errorMessage = "Something went wrong";
    const state = Reducer(null, Actions.getUsersReceiveErrorActionCreator(errorMessage));
    expect(state).toMatchSnapshot();
  })
});
