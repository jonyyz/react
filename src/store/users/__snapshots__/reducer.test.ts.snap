// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users reducer UsersReceiveAction 1`] = `
Object {
  "fetching": true,
}
`;

exports[`Users reducer UsersReceiveAction 2`] = `
Object {
  "fetched": true,
  "fetching": false,
  "users": Array [
    Object {
      "city": "Walla Walla Washington",
      "companyName": "ACME",
      "id": 42,
      "name": "Mel Blanc",
      "website": "acme.org",
    },
  ],
}
`;

exports[`Users reducer UsersReceiveError 1`] = `
Object {
  "fetchError": "Something went wrong",
  "fetching": false,
}
`;
